packages/core/strapi/lib/types/core/common/uid.d.ts

```diff
@@ -41,22 +41,18 @@ export type Schema = Registry.Keys<
   UID.ContentType | UID.Component<ComponentCategory>
 >;
 
-export type IsCollectionType<TContentTypeUID extends ContentType> = Utils.Expression.Extends<
-  Shared.ContentTypes[TContentTypeUID],
-  SchemaNamespace.CollectionType
->;
+export type IsCollectionType<TSchemaUID extends Schema> = TSchemaUID extends CollectionType
```

Comment:
I need some kind of explanation for what is happening here :D I see TSchemaUID extends Schema ? shouldn't it be SchemaUID or sth like that ?  same extends CollectionType or CollectionTypeUID ? I don't know if we are looking at shcemas or just their uids anymore. 

Can you explain what the expression on line 45 does ? If we know the UID is part of the collection types uids why do we need to check again the schema ?

Reply:
What do you think about this?
(I'm a bit worried about the double Schema type and possible conflicts)
```ts
import type { Shared, Schema, Common, Registry, UID, Utils } from '@strapi/strapi';

export type Service = Registry.Keys<Shared.Services, UID.Service>;

export type Controller = Registry.Keys<Shared.Controllers, UID.Controller>;

export type Policy = Registry.Keys<Shared.Policies, UID.Policy>;

export type Middleware = Registry.Keys<Shared.Middlewares, UID.Middleware>;

export type ContentType = Registry.Keys<Shared.ContentTypes, UID.ContentType>;

export type CollectionType = Utils.Guard.Never<
  // extract uids only for collection types
  Extract<Utils.Object.KeysBy<Shared.ContentTypes, Schema.CollectionType>, ContentType>,
  // if no collection type is found (never), fallback to a generic content type uid
  ContentType
>;

export type SingleType = Utils.Guard.Never<
  // extract uids only for single types
  Extract<Utils.Object.KeysBy<Shared.ContentTypes, Schema.SingleType>, ContentType>,
  // if no single type is found (never), fallback to a generic content type uid
  ContentType
>;

export type Component = Registry.Keys<Shared.Components, UID.Component>;
export type ComponentCategory = Component extends UID.Component<infer TCategory>
  ? TCategory
  : never;

export type Schema = Registry.Keys<
  Common.Schemas,
  UID.ContentType | UID.Component<ComponentCategory>
>;

export type IsCollectionType<TSchemaUID extends Common.UID.Schema> =
  TSchemaUID extends CollectionType
    ? Utils.Expression.Extends<Shared.ContentTypes[TSchemaUID], Schema.CollectionType>
    : Utils.Expression.False;

export type IsSingleType<TSchemaUID extends Common.UID.Schema> = TSchemaUID extends SingleType
  ? Utils.Expression.Extends<Shared.ContentTypes[TSchemaUID], Schema.SingleType>
  : Utils.Expression.False;

export type IsComponent<TSchemaUID extends Common.UID.Schema> = TSchemaUID extends Component
  ? Utils.Expression.Extends<Shared.Components[TSchemaUID], Schema.Component>
  : Utils.Expression.False;

export type IsContentType<TSchemaUID extends Common.UID.Schema> = TSchemaUID extends ContentType
  ? Utils.Expression.Or<IsCollectionType<TSchemaUID>, IsSingleType<TSchemaUID>>
  : Utils.Expression.False;
```

Reply:
> Can you explain what the expression on line 45 does ? If we know the UID is part of the collection types uids why do we need to check again the schema ?

The test on the uid is to make sure the given uid matches the format, but multiple UIDs formats can overlap (single-type/collection-types). The test on the schema makes sure that we're testing the actual type by discriminating on schema values (schema.modelType or other properties)

Reply:
I'll resolve it for now so we address this more globally on the org of the types and the public interfaces :)